### ISSUE TITLE
Q: check if assembly is compiled AOT

### ISSUE THREAD

**Issue by tmds** [dotnet/runtime#1802]
What is a way to check on Linux if an assembly was compiled AOT, and for what target?

cc @jkotas

**Comment by jkotas** [comment]
- Open the file using `System.Reflection.Metadata.MetadataReader`
- Non-null `CorHeader.ManagedNativeHeaderDirectory` means that the assembly was AOT compiled
- `CoffHeader.Machine` is the target machine (e.g. `Machine.Arm64`). For Unix platforms, the target machine is xored with (`IMAGE_FILE_MACHINE_NATIVE_OS_OVERRIDE`)[https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/coreclr/src/inc/pedecoder.h#L90]. (You will have to copy these constants. They are not part of System.Reflection.Metadata.)

@dotnet/crossgen-contrib Do we happen to have a code for this somewhere that can be copy-pasted?

**Comment by jkotas** [comment]
https://github.com/dotnet/runtime/blob/0812d49aeb7ae83bf04a6799f47594ad8a47cfdd/src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs#L59 is the managed definition of `IMAGE_FILE_MACHINE_NATIVE_OS_OVERRIDE` that you can copy.

**Comment by cshung** [comment]
`ILCompiler.Reflection.ReadyToRun.ReadyToRunReader..ctor()` will throw `BadImageFormatException` if it is not a ReadyToRun image.

The `Machine` and `OperatingSystem` property can be used to determine the target.

**Comment by trylek** [comment]
I'm afraid no copy-paste code exists now. As Andrew pointed out, R2RDump is the only place we're checking input files now - Crossgen2 ignores OS / architecture for input MSIL (which is likely an omission) - basically looping over all the enum values as I don't see any other way how to decipher the XORed value. The relevant bit is here:

https://github.com/dotnet/runtime/blob/247c5b96a226c3b1bd5ecca8a139b7c7da97d5cf/src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs#L365

**Comment by cshung** [comment]
Why would one want to copy-paste code when you can just reference `ILCompiler.Reflection.ReadyToRun.dll` and use it? It is currently targetting `.NET Standard 2.0` and is usable in many places.

In fact, that's the whole point of the refactoring, make the ready to run image parsing logic reusable.

**Comment by trylek** [comment]
Well, I guess this holds for projects within dotnet/runtime but I suspect that arbitrarily referencing ILC.R.R2R.dll from a 3rd party project might be a pain in the a$$.

**Comment by cshung** [comment]
It is not all that hard, one clone and builds the tree, creates a package, hosts it on a package feed. Once this is done, any project can reference it.

Alternatively, I have already created a package [here](https://github.com/icsharpcode/ILSpy/blob/master/NuGet.config) for ILSpy, feel free to reuse the binary for any purpose.

**Comment by tmds** [comment]
@jkotas @trylek @cshung thank you for this info, I can figure out now what assemblies are r2r compiled.

> Alternatively, I have already created a package here for ILSpy, feel free to reuse the binary for any purpose.

@cshung will the next version of ilspy be able to show this info?

**Comment by cshung** [comment]
> @cshung will the next version of ILSpy be able to show this info?

Yes, it will be available. I hesitated to reply because it was a work in progress. The [PR](https://github.com/icsharpcode/ILSpy/pull/1914) was merged yesterday!

In particular, you will be able to see the machine (x86/amd64/...), operating system (windows/linux/...), compiler identifier (e.g. crossgen/crossgen2), and the disassembly of individual methods.
